otaclient/app/create_standby/common.py

```diff
@@ -303,85 +298,164 @@ def _cal_and_prepare_old_slot_delta(self):
+                    # if delta_src meta is available, only process the file listed
+                    # in the delta_src meta, otherwise move it to rm_list
+                    if self._delta_src_path_hash_dict:
+                        if _hash := self._delta_src_path_hash_dict.get(
+                            str(canonical_fpath)
+                        ):
+                            futs.append(
+                                pool.submit(
+                                    self._process_file_in_old_slot,
+                                    delta_src_fpath,
+                                    _hash=_hash,
+                                )
+                            )
+                        else:
+                            self._rm.append(str(canonical_fpath))
+                    # delta_src meta is not available,
+                    # directly hash and collect the file
+                    else:
+                        futs.append(
+                            pool.submit(self._process_file_in_old_slot, delta_src_fpath)
+                        )
+                # wait for all files under this dir to be finished
+                logger.debug(f"{delta_src_curdir_path=} done")
+                wait(futs)
+
+        # calculate the files list that we should download from remote
+        # the hash in the self._hash_set after local delta preparation representing
+        # the file we need to download from remote
+        for _hash, _reginf_set in self._new.items():
+            if _hash in self._new_hash_list:
+                # pick one entry from the reginf set for downloading
+                _iter = _reginf_set.iter_entries()
+                _, _entry = next(_iter)
+                self._download_list.append(_entry)
+                self.total_download_files_size += _entry.size if _entry.size else 0
+        self._new_hash_list.clear()
+
+    # API
+
+    def calculate_and_process_delta(
+        self,
+        *,
+        delta_src_reg: Path,  # path to delta_src regulars.txt
+        new_reg: Path,  # path to new image regulars.txt
+        new_dirs: Path,  # path to dirs.txt
+    ) -> DeltaBundle:
+        with open(new_dirs, "r") as f:
+            for _dir in map(DirectoryInf, f):
+                self._new_dirs[_dir] = None
+        # pre-load from new regulars.txt
+        with open(new_reg, "r") as f:
+            for _entry in map(RegularInf.parse_reginf, f):
+                self.total_regulars_num += 1
+                self._new.add_entry(_entry)
+                self._new_hash_list.add(_entry.sha256hash)
+        self._stats_collector.store.total_regular_files = self.total_regulars_num
+        # pre-load delta src's reginf if presented
+        if Path(delta_src_reg).is_file():
+            with open(delta_src_reg, "r") as f:
+                for _entry in map(RegularInf.parse_reginf, f):
+                    self._delta_src_path_hash_dict[str(_entry.path)] = _entry.sha256hash
```

Comment:
I think this is the new feature. It should be the another PR. Is it well tested?
This feature changes the ota client behavior if `/opt/ota/image-meta/regulars.txt` exists or not.

Reply:
@obi-t4  Yes, it can speed up the delta generating process if the current running slot has OTA image meta files saved in `/opt/ota/image-meta`. 
That is actually the leftover TODO for `DeltaGenerator` implementation. If needed I can remove it and re-add it in another PR.

Reply:
We need to do two test cases:
1. /opt/ota/image-meta/regulars.txt exists
2. /opt/ota/image-meta/regulars.txt doesn't exist
In this PR, I'd like to focus on the new retry-strategy.